### ISSUE TITLE
ENH: fft/signal: helper functions for converting between coefficients and power/phase

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#18319]
### Describe the solution you'd like.

One of the suggestions in https://github.com/scipy/scipy/issues/16191 was to add "helper functions within [scipy.fft] that 1) converts the fft coefficients to power (V**2/Hz) and 2) converts power and phase to fft coefficients".

### Describe alternatives you've considered.

Not adding such a function (sometimes we avoid simple convenience functions).
Adding options to do the conversion within existing fft or signal functions (more work, probably not easier to use).

*Initial conversation on the mailing list is [here](https://mail.python.org/archives/list/scipy-dev@python.org/thread/O3BZ33K7SAALUO3DE5UNL45QB72H3GC5/).*

**Comment by rohan472000** [comment]
The `fft_to_power` function takes an array of FFT coefficients and the sampling rate as inputs, and returns an array of power spectrum values in V**2/Hz. The `power_phase_to_fft` function takes an array of power spectrum values and an array of phase spectrum values as inputs, and returns an array of FFT coefficients.

These helper functions can be used in conjunction with existing FFT and signal processing functions in `scipy.fft` to perform various signal processing tasks, such as filtering, spectral analysis, and synthesis.


    import numpy as np
    
    def fft_to_power(fft_array, sampling_rate):
        """
        Convert FFT coefficients to power (V**2/Hz).
    
        Parameters:
            fft_array (ndarray): FFT coefficients.
            sampling_rate (float): Sampling rate in Hz.
    
        Returns:
            ndarray: Power spectrum in V**2/Hz.
        """
        power_spectrum = (np.abs(fft_array)**2)/sampling_rate
        return power_spectrum
    
    
    def power_phase_to_fft(power_spectrum, phase_spectrum):
        """
        Convert power (V**2/Hz) and phase to FFT coefficients.
    
        Parameters:
            power_spectrum (ndarray): Power spectrum in V**2/Hz.
            phase_spectrum (ndarray): Phase spectrum in radians.
    
        Returns:
            ndarray: FFT coefficients.
        """
        complex_spectrum = np.sqrt(power_spectrum) * np.exp(1j * phase_spectrum)
        fft_array = np.fft.irfft(complex_spectrum)
        return fft_array

**Comment by ilayn** [comment]
We have to note that FFT is performed in many disciplines and this is a very electrical engineering specific wording. Especially units are quite foreign to, non-EE or even control engineering people. It's a rite of passage in the early days that you force your brain into dB usage. 9 out of 10 people can't explain what V/sqrt(Hz), root power etc is. 

Therefore I am of the opinion that we should keep these functions lingo-free as much as possible. Also I'm not sure that we need these "two liners" are needed as a public API functions.

**Comment by rohan472000** [comment]
Thank you for bringing up this point. Indeed, using technical terms and units that are specific to a certain discipline may cause confusion for users from other disciplines.

To address this concern, we can modify the function names and docstrings to use more generic terms that are understandable by a wider audience. For example, instead of `fft_to_power`, we can use `spectrum_to_power` or `coefficients_to_power_spectrum`. Similarly, instead of `power_phase_to_fft`, we can use `power_spectrum_and_phase_to_coefficients` or `complex_spectrum_to_coefficients`.

As for the necessity of these functions, it depends on the use cases and the level of abstraction that the library aims to provide. If the library is meant to be a high-level interface for signal processing tasks, then these functions might be helpful to simplify the workflow and make the library more accessible to users who are not familiar with the underlying mathematics. However, if the library is intended for advanced users who prefer to work with low-level building blocks, then these functions might be redundant.

Afterall, the decision of whether to include these functions in the public API should be based on a thorough evaluation of the library's target audience, use cases, and design principles.

**Comment by vollbier** [comment]
In my open PR #17408 for improving the STFT, I also dealt with scaling FFT coefficients. I found the topic to a be bit more intricate than @rohan472000's proposal suggests: 

In spectral analysis, typically the input signal is multiplied with a windowing function to suppress spectral leakage to some degree. When the integral over the windowing function is one, the FFT coefficients can be interpreted as the coefficients of a Fourier series made up of windowed complex exponents. I.e., the coefficients have the same unit `V `and the correct scaling. When plotting the absolute values, this scaling is often referred to as a "magnitude spectrum". When the integral over the absolute square of the windowing function is one, than the FFT can be interpreted as a "power spectral density" (PSD), i.e. having unit `V / sqrt(Hz)`.

Not explicitly considering the windowing function properties causes some complications:

1. Due to convention, the windowing function in [scipy.signal.windows][1] module are not normalized.
2. When using no window, i.e., using an implicit [rectangular window][2], one needs to divide the coefficients by the number of input samples to obtain a magnitude spectrum. This scaling is assumed for the input of `fft_to_power` function, if I interpret it correctly.
3. Determining the signal's power of a given frequency interval by numerically integrating over the corresponding PSD interval is a typical application. When using a one-sided spectrum (i.e., [rfft][3]), the coefficients needs to be doubled, but not the 0 Hz bin. Furthermore, if the number of frequency bins is even, the last bin should not be doubled either. I think, a `fft_to_power` function should account for this use case as well.

Scaling is also a problem with the STFT, as illustrated by #14093, #1600, #14740, and #15963.
In my PR #17408 (reviews are very welcome!), I am addressing this topic by explicitly providing the scaling factors. An example for the intricacies is this [table][4], which shows the scaling correspondences between the existing and my new STFT implementation.

In summary, I agree with @ilayn that these "two-liners" are probably not that useful on their own. I tried to show with my lengthy remark that the context of the application should not be neglected. Perhaps an extensive example in the documentation, which discusses the typical use-cases, might be more beneficial to the average user.


[1]: https://docs.scipy.org/doc/scipy/reference/signal.windows.html
[2]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.windows.boxcar.html
[3]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.rfft.html
[4]: https://github.com/scipy/scipy/pull/17408/files#diff-76d0be0c96d39f6d09fdf65a7432d5faf82739a764dabd9557ade723d42b072dR1595

**Comment by mdhaber** [comment]
Looks like there are some solid reasons not to add functions for these features, so I'll go ahead and close this issue, but an example in the tutorial sounds great. I do not have enough experience to write the example myself, but I'd be happy to review one if you @vollbier or @rohan472000 would be willing to give it a shot. Thank you all for your thoughts!
